crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4640,10 +4650,264 @@ mod tests {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                x = 1
+            except NameError:
+                x = 2
+            except TypeError:
+                pass
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        // `x` may be unbound if the `except TypeError` branch is taken;
+        // it is bound if the the `try` block runs to completion or the `except NameError` branch is taken:
+        assert_public_ty(&db, "src/a.py", "x", "Unbound | Literal[1, 2]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn except_handler_definitions_with_pre_existing_definitions() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            e = 1
+            f = 2
+
+            try:
+                x
+            except NameError:
+                e = 42
+            except TypeError:
+                f = 43
+            except RuntimeError:
+                f = 44
+            except OSError:
+                g = 45
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+        assert_public_ty(&db, "src/a.py", "e", "Literal[1, 42]");
+        assert_public_ty(&db, "src/a.py", "f", "Literal[2, 43, 44]");
+        assert_public_ty(&db, "src/a.py", "g", "Unbound | Literal[45]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn except_handler_definitions_with_finally_branch() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                x
+            except NameError:
+                e = 42
+            finally:
+                e = 44
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        // The `finally` block is always executed:
+        assert_public_ty(&db, "src/a.py", "e", "Literal[44]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn except_handler_definitions_with_else_branch() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                x = 42
+            except NameError:
+                x = 43
+            except TypeError:
+                x = 44
+            else:
+                x = 45
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        // Either one of the `except` branches or the `else` branch must have been taken;
+        // the `x` definition in the `try` block must have been overridden after completion
+        // of the entire `try`/`except` block:
+        assert_public_ty(&db, "src/a.py", "x", "Literal[43, 44, 45]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn except_handler_multiple_definitions_in_try_block() -> anyhow::Result<()> {
```

Comment:
I think a test that is missing here is something verifying that _within_ the `else` block, names bound in the `try` block are definitely bound. (This is along the same lines as the above comment; we're generally weak on tests demonstrating inferred types inside a particular block, rather than at end of scope.)

But as mentioned above, this is difficult to do with current test infra. (You can assign the name to another name, but then that other name is still potentially unbound, so the test isn't super clear.) But it would be nice to have a TODO to add this test, once we have the new test framework.